File: commands/stats/dominance.py
# https://github.com/viktorexe/terristats-discord-bot
import discord
from discord.ext import commands
from discord import app_commands
from collections import Counter
from database import db

class Dominance(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    async def map_autocomplete(self, interaction: discord.Interaction, current: str):
        try:
            maps = await db.db.clanwins.distinct('map')
            maps = [m for m in maps if m and isinstance(m, str)]
            maps.sort()
            if current:
                filtered_maps = [m for m in maps if current.lower() in m.lower()]
            else:
                filtered_maps = maps
            return [app_commands.Choice(name=map_name, value=map_name) for map_name in filtered_maps[:25]]
        except:
            return []

    @app_commands.command(name='dominance', description='Show clan dominance on specific map by percentage')
    @app_commands.autocomplete(map=map_autocomplete)
    async def dominance(self, interaction: discord.Interaction, map: str):
        try:
            await interaction.response.defer()
            
            query = {'map': {'$regex': f'^{map.strip()}$', '$options': 'i'}}
            
            pipeline = [
                {'$match': query},
                {'$facet': {
                    'stats': [{'$group': {
                        '_id': None,
                        'total': {'$sum': 1},
                        'points': {'$sum': '$player_count'}
                    }}],
                    'clans': [{'$group': {'_id': '$clan_name', 'count': {'$sum': 1}}}, {'$sort': {'count': -1}}, {'$limit': 10}]
                }}
            ]
            
            result = await db.db.clanwins.aggregate(pipeline).to_list(1)
            if not result or not result[0].get('stats'):
                await interaction.followup.send(embed=discord.Embed(
                    title=f"{map} Dominance", description="No games found.", color=0xff0000
                ))
                return
            
            data = result[0]
            stats = data['stats'][0] if data['stats'] else {}
            total = stats.get('total', 0)
            total_points = stats.get('points', 0)
            top_clans = [(c['_id'], c['count']) for c in data.get('clans', [])]
            
            embed = discord.Embed(title=f"{map} Dominance", description=f"{total} games | {total_points} points", color=0x2b2d31)
            
            rankings = []
            for i, (clan, count) in enumerate(top_clans, 1):
                percentage = (count / total) * 100
                bar_length = int(percentage / 10)
                bar = '█' * bar_length + '░' * (10 - bar_length)
                rankings.append(f"#{i} [{clan}]\n{bar} {percentage:.1f}% ({count})")
            
            embed.add_field(name="Clan Dominance", value='\n\n'.join(rankings), inline=False)
            
            total_games = await db.db.clanwins.count_documents({})
            embed.set_footer(text=f"Data from 18 Nov 2025 | {total_games:,} games tracked")
            await interaction.followup.send(embed=embed)
            
        except Exception as e:
            print(f"Error in dominance: {e}")
            try:
                await interaction.followup.send(embed=discord.Embed(
                    title="Error", description="Failed to calculate dominance.", color=0xff0000
                ))
            except:
                pass

async def setup(bot):
    await bot.add_cog(Dominance(bot))

File: commands/stats/dominance.js
// https://github.com/viktorexe/terristats-discord-bot
// /dominance - clan dominance on a specific map, shown as percentage bars.
const { SlashCommandBuilder, EmbedBuilder } = require('discord.js');
const db = require('../../database');
const { gamesFooter } = require('../../utils');

module.exports = {
  data: new SlashCommandBuilder()
    .setName('dominance')
    .setDescription('Show clan dominance on specific map by percentage')
    .addStringOption((o) =>
      o
        .setName('map')
        .setDescription('Select a map')
        .setRequired(true)
        .setAutocomplete(true),
    ),

  // Suggest up to 25 map names matching what the user has typed so far.
  async autocomplete(interaction) {
    try {
      const current = interaction.options.getFocused().toLowerCase();
      let maps = await db.collection('clanwins').distinct('map');
      maps = maps.filter((m) => m && typeof m === 'string').sort();
      const filtered = current
        ? maps.filter((m) => m.toLowerCase().includes(current))
        : maps;
      await interaction.respond(
        filtered.slice(0, 25).map((m) => ({ name: m, value: m })),
      );
    } catch (err) {
      await interaction.respond([]);
    }
  },

  async execute(interaction) {
    try {
      await interaction.deferReply();

      const map = interaction.options.getString('map');
      const query = { map: { $regex: `^${map.trim()}$`, $options: 'i' } };

      const pipeline = [
        { $match: query },
        {
          $facet: {
            stats: [
              { $group: { _id: null, total: { $sum: 1 }, points: { $sum: '$player_count' } } },
            ],
            clans: [
              { $group: { _id: '$clan_name', count: { $sum: 1 } } },
              { $sort: { count: -1 } },
              { $limit: 10 },
            ],
          },
        },
      ];

      const result = await db.collection('clanwins').aggregate(pipeline).toArray();
      if (!result.length || !result[0].stats.length) {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle(`${map} Dominance`)
              .setDescription('No games found.')
              .setColor(0xff0000),
          ],
        });
        return;
      }

      const data = result[0];
      const stats = data.stats[0] || {};
      const total = stats.total || 0;
      const totalPoints = stats.points || 0;
      const topClans = data.clans.map((c) => [c._id, c.count]);

      const embed = new EmbedBuilder()
        .setTitle(`${map} Dominance`)
        .setDescription(`${total} games | ${totalPoints} points`)
        .setColor(0x2b2d31);

      const rankings = topClans.map(([clan, count], i) => {
        const percentage = (count / total) * 100;
        const barLength = Math.floor(percentage / 10);
        const bar = '█'.repeat(barLength) + '░'.repeat(10 - barLength);
        return `#${i + 1} [${clan}]\n${bar} ${percentage.toFixed(1)}% (${count})`;
      });
      embed.addFields({ name: 'Clan Dominance', value: rankings.join('\n\n'), inline: false });

      const totalGames = await db.collection('clanwins').countDocuments({});
      embed.setFooter({ text: gamesFooter(totalGames) });
      await interaction.editReply({ embeds: [embed] });
    } catch (err) {
      console.log(`Error in dominance: ${err.message}`);
      try {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle('Error')
              .setDescription('Failed to calculate dominance.')
              .setColor(0xff0000),
          ],
        });
      } catch (e) {
        // ignore
      }
    }
  },
};
